libazureinit/src/kvp.rs

```diff
@@ -0,0 +1,649 @@
+
+                let start_time_dt = DateTime::<Utc>::from(
+                    UNIX_EPOCH
+                        + start_time
+                            .duration_since(UNIX_EPOCH)
+                            .unwrap_or_default(),
+                )
+                .format("%Y-%m-%dT%H:%M:%S%.3fZ");
+
+                let end_time_dt = DateTime::<Utc>::from(
+                    UNIX_EPOCH
+                        + end_time
+                            .duration_since(UNIX_EPOCH)
+                            .unwrap_or_default(),
+                )
+                .format("%Y-%m-%dT%H:%M:%S%.3fZ");
+
+                let event_value = format!(
+                    "Start: {} | End: {} | Status: {}",
+                    start_time_dt, end_time_dt, span_status
+                );
+
+                handle_kvp_operation(
+                    self,
+                    span_status.level(),
+                    span_context.name(),
+                    &span_id.to_string(),
+                    &event_value,
+                );
+            }
+        }
+    }
+}
+
+/// Handles the orchestration of key-value pair (KVP) encoding and logging operations
+/// by generating a unique event key, encoding it with the provided value, and sending
+/// it to the `EmitKVPLayer` for logging.
+pub fn handle_kvp_operation(
+    emit_kvp_layer: &EmitKVPLayer,
+    event_level: &str,
+    event_name: &str,
+    span_id: &str,
+    event_value: &str,
+) {
+    let event_key = generate_event_key(event_level, event_name, span_id);
+    let encoded_kvp = encode_kvp_item(&event_key, event_value);
+    let encoded_kvp_flattened: Vec<u8> = encoded_kvp.concat();
+    emit_kvp_layer.send_event(encoded_kvp_flattened);
+}
+
+/// Generates a unique event key by combining the event level, name, and span ID.
+///
+/// # Arguments
+/// * `event_level` - The logging level (e.g., "INFO", "DEBUG").
+/// * `event_name` - The name of the event.
+/// * `span_id` - A unique identifier for the span.
+fn generate_event_key(
+    event_level: &str,
+    event_name: &str,
+    span_id: &str,
+) -> String {
+    format!(
+        "{}|{}|{}|{}",
+        EVENT_PREFIX, event_level, event_name, span_id
+    )
+}
+
+/// Encodes a key-value pair (KVP) into one or more byte slices. If the value
+/// exceeds the allowed size, it is split into multiple slices for encoding.
+/// This is used for logging events to a KVP file.
+///
+/// # Arguments
+/// * `key` - The key as a string slice.
+/// * `value` - The value associated with the key.
+fn encode_kvp_item(key: &str, value: &str) -> Vec<Vec<u8>> {
+    let key_bytes = key.as_bytes();
+    let value_bytes = value.as_bytes();
+
+    let key_len = key_bytes.len().min(HV_KVP_EXCHANGE_MAX_KEY_SIZE);
+    let mut key_buf = vec![0u8; HV_KVP_EXCHANGE_MAX_KEY_SIZE];
+    key_buf[..key_len].copy_from_slice(&key_bytes[..key_len]);
+
+    if value_bytes.len() <= HV_KVP_AZURE_MAX_VALUE_SIZE {
```

Comment:
Is there a unit test that would validate that the slice's value's length is no more than the azure's limit?

Reply:
I didn't have a unit test for it, just the general check that it's below HV_KVP_AZURE_MAX_VALUE_SIZE, but there is a unit test now -- the last test at the end of the test block called `test_encode_kvp_item_value_length`. It creates an oversized KVP value and then, in checking that the value is too large, slices it, and then checks that each slice’s value length adheres to the Azure-defined limit.